From: Veera Datla <veera.datla@digi.com>
Date: Wed, 18 Mar 2020 14:01:02 +1000
Subject: modem: fix registered status when connected

ModemManager status is showing as Registered while it is connected.

ModemManager state is changing from CONNECTED to ENABLED state
when the modem is idle.

Introduced a check in ModemManager to see if previous state
is CONNECTED and any bearer is present, if so state is updated
to previous state (CONNECTED) or else continue with the state
change.
---
 src/mm-iface-modem.c | 19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/src/mm-iface-modem.c b/src/mm-iface-modem.c
index 0c2f9bc8..5ec495a8 100644
--- a/src/mm-iface-modem.c
+++ b/src/mm-iface-modem.c
@@ -1524,8 +1524,23 @@ __iface_modem_update_state_internal (MMIfaceModem *self,
     }
 
     /* Enabled may really be searching or registered */
-    if (new_state == MM_MODEM_STATE_ENABLED)
-        new_state = get_consolidated_subsystem_state (self);
+    if (new_state == MM_MODEM_STATE_ENABLED) {
+        guint connected = 0;
+        if (old_state == MM_MODEM_STATE_CONNECTED) {
+            mm_bearer_list_foreach (bearer_list,
+                                    (MMBearerListForeachFunc)bearer_list_count_connected,
+                                    &connected);
+            mm_obj_info (self, "Bearer connected = %d, when trying to change state (%s -> %s)",
+                     connected,
+                     mm_modem_state_get_string (old_state),
+                     mm_modem_state_get_string (new_state));
+        }
+        /* Don't change state */
+        if (connected > 0)
+            new_state = old_state;
+        else
+            new_state = get_consolidated_subsystem_state (self);
+    }
 
     /* Update state only if different */
     if (new_state != old_state) {
